### ISSUE TITLE
Clang Kit scanning will accidentally use GCC for C++

### ISSUE THREAD

**Issue by ProfFan** [microsoft/vscode-cmake-tools#1759]
### Brief Issue Summary

The Clang kit scan is broken with current LLVM 11. It will show as

![image](https://user-images.githubusercontent.com/5358374/114488313-2672c480-9bdf-11eb-99a3-8c78331ba30a.png)

and then call CMake with only the `CMAKE_C_COMPILER` set. This leads to the mixing of GCC and Clang toolchains

![image](https://user-images.githubusercontent.com/5358374/114488583-9bde9500-9bdf-11eb-9315-934c0f962746.png)

This is very subtle and causes a lot of confusion.

### Expected:

1. Kits without a C++ compiler should not show if the project is C++. Or such kits should have a different style to warn users.
2. LLVM 11.1 Kits should be properly detected

### CMake Tools Log

```
[kit] Found Kit: GCC for aarch64-linux-gnu 10.2.0
[kit] Found Kit: GCC for aarch64-linux-gnu 10.2.0
[kit] Found Kit: GCC for arm-none-eabi 10.2.0
[kit] Found Kit: GCC for arm-none-eabi 10.2.0
[kit] Found Kit: Clang 11.1.0
[kit] Found Kit: Clang 11.1.0
[kit] Found Kit: GCC 10.2.0
[kit] Found Kit: GCC 7.4.1
[kit] Found Kit: GCC 8.4.0
[kit] Found Kit: GCC 9.3.0
[kit] Found Kit: GCC for riscv64-linux-gnu 10.2.0
[kit] Found Kit: GCC for riscv64-linux-gnu 10.2.0
[kit] Found Kit: GCC for riscv64-unknown-elf 8.2.0
[kit] Found Kit: GCC for riscv64-unknown-elf 8.2.0
[kit] Found Kit: GCC for x86_64-pc-linux-gnu 10.2.0
[kit] Found Kit: GCC for x86_64-pc-linux-gnu 10.2.0
[kit] Found Kit: GCC for x86_64-pc-linux-gnu 7.4.1
[kit] Found Kit: GCC for x86_64-pc-linux-gnu 7.4.1
[kit] Found Kit: GCC for x86_64-pc-linux-gnu 8.4.0
[kit] Found Kit: GCC for x86_64-pc-linux-gnu 8.4.0
[kit] Found Kit: GCC for x86_64-pc-linux-gnu 9.3.0
[kit] Found Kit: GCC for x86_64-pc-linux-gnu 9.3.0
[kit] Found Kit: GCC 10.2.0
[kit] Successfully loaded 15 kits from /home/fan/.local/share/CMakeTools/cmake-tools-kits.json
```

### Platform and Versions

- **Operating System**: Arch Linux
- **CMake Version**: 3.20.1
- **VSCode Version**: 1.55.1
- **CMake Tools Extension Version**: 1.6.0
- **Compiler/Toolchain**: LLVM 11.1

**Comment by bobbrow** [comment]
Thank you for reporting this.  We may not get to it immediately, but we are always open to accept a PR if someone has a chance to investigate this before we do.

**Comment by andreeis** [comment]
Your first suggestion (to not show kits that are missing CXX compiler when being within a C++ project) sounds useful and we will consider it in the future.
Regarding the second expectation (to detect a complete clang 11.1.0 kit), can you confirm if you have clang++-11 in /usr/bin? If not, maybe your installation of clang 11.1 is corrupt and you can try a reinstall. If yes, can you copy paste the output from "clang++-11 --version" and "clang-11 --version"?

**Comment by ProfFan** [comment]
Sure!

```
> /usr/bin/clang++ --version
clang version 11.1.0
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
> /usr/bin/clang --version
clang version 11.1.0
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

**Comment by andreeis** [comment]
Ok, weird. How does the llvm kit looks to be written in cmake-tools-kits.json? To open this kits file and copy paste the kit definition for us, you can use the command "CMake: Edit User-Local CMake Kits" in the palette.
Also, worth trying, delete the LLVM kit from that file and run the command "CMake: Scan for kits" and see how the LLVM 11.1 kit definition is written (probably still wrong).
Then, until we figure out how to fix this, to get unblocked you can write the missing CXX line yourself in the LLVM 11.1 kit definition in the cmake-tools-kits.json file.
After getting more info from you, we may split this report in 2: one about not including a kit that does not have CXX compiler when we detect a C++ project and the other about why we don't detect the ++ portion of the clang llvm 11.1 kit.

**Comment by ProfFan** [comment]
I tried removing and rescanning, it's the same result.

The JSON file (sorry I have too many compilers :) )
```[
  {
    "name": "Clang 11.1.0",
    "compilers": {
      "C": "/usr/bin/clang-11"
    }
  },
  {
    "name": "Clang 11.1.0 x86_64-pc-linux-gnu",
    "compilers": {
      "C": "/usr/bin/clang-11"
    }
  },
  {
    "name": "GCC 10.1.0",
    "compilers": {
      "C": "/bin/gcc",
      "CXX": "/bin/g++"
    }
  },
  {
    "name": "GCC 10.2.0",
    "compilers": {
      "C": "/opt/cuda/bin/gcc",
      "CXX": "/opt/cuda/bin/g++"
    }
  },
  {
    "name": "GCC 10.2.0 riscv64-linux-gnu",
    "compilers": {
      "C": "/usr/bin/riscv64-linux-gnu-gcc-10.2.0"
    }
  },
  {
    "name": "GCC 10.2.0 x86_64-pc-linux-gnu",
    "compilers": {
      "C": "/opt/cuda/bin/gcc",
      "CXX": "/opt/cuda/bin/g++"
    }
  },
  {
    "name": "GCC 10.3.0 aarch64-linux-gnu",
    "compilers": {
      "C": "/usr/bin/aarch64-linux-gnu-gcc-10.3.0"
    }
  },
  {
    "name": "GCC 10.3.0 arm-none-eabi",
    "compilers": {
      "C": "/usr/bin/arm-none-eabi-gcc-10.3.0"
    }
  },
  {
    "name": "GCC 7.4.1",
    "compilers": {
      "C": "/usr/bin/gcc-7",
      "CXX": "/usr/bin/g++-7"
    }
  },
  {
    "name": "GCC 7.4.1 x86_64-pc-linux-gnu",
    "compilers": {
      "C": "/usr/bin/x86_64-pc-linux-gnu-gcc-7.4.1"
    }
  },
  {
    "name": "GCC 8.2.0 riscv64-unknown-elf",
    "compilers": {
      "C": "/usr/bin/riscv64-unknown-elf-gcc-8.2.0"
    }
  },
  {
    "name": "GCC 8.2.1",
    "compilers": {
      "C": "/bin/gcc",
      "CXX": "/bin/g++"
    }
  },
  {
    "name": "GCC 8.4.0",
    "compilers": {
      "C": "/usr/bin/gcc-8",
      "CXX": "/usr/bin/g++-8"
    }
  },
  {
    "name": "GCC 8.4.0 x86_64-pc-linux-gnu",
    "compilers": {
      "C": "/usr/bin/x86_64-pc-linux-gnu-gcc-8.4.0"
    }
  },
  {
    "name": "GCC 9.3.0",
    "compilers": {
      "C": "/usr/bin/gcc-9",
      "CXX": "/usr/bin/g++-9"
    }
  },
  {
    "name": "GCC 9.3.0 x86_64-pc-linux-gnu",
    "compilers": {
      "C": "/usr/bin/x86_64-pc-linux-gnu-gcc-9.3.0"
    }
  },
  {
    "name": "GCC for riscv64-linux-gnu 10.2.0",
    "compilers": {
      "C": "/usr/bin/riscv64-linux-gnu-gcc-10.2.0"
    }
  },
  {
    "name": "GCC for riscv64-unknown-elf 8.2.0",
    "compilers": {
      "C": "/usr/bin/riscv64-unknown-elf-gcc-8.2.0"
    }
  },
  {
    "name": "GCC for x86_64-pc-linux-gnu 10.2.0",
    "compilers": {
      "C": "/usr/bin/x86_64-pc-linux-gnu-gcc-10.2.0"
    }
  },
  {
    "name": "GCC for x86_64-pc-linux-gnu 7.4.1",
    "compilers": {
      "C": "/usr/bin/x86_64-pc-linux-gnu-gcc-7.4.1"
    }
  },
  {
    "name": "GCC for x86_64-pc-linux-gnu 8.4.0",
    "compilers": {
      "C": "/usr/bin/x86_64-pc-linux-gnu-gcc-8.4.0"
    }
  },
  {
    "name": "GCC for x86_64-pc-linux-gnu 9.3.0",
    "compilers": {
      "C": "/usr/bin/x86_64-pc-linux-gnu-gcc-9.3.0"
    }
  }
]
```

**Comment by andreeis** [comment]
Ok I'll install llvm 11.1 on one of my ubuntus and I hope to reproduce this.

**Comment by minhducsun2002** [comment]
Can confirm identical behaviour with VS Code 1.55.2, CMake Tools 1.7.1, and similar LLVM & CMake version on Arch Linux.

**Comment by riri** [comment]
I get the same, clang++ is not detected and g++ is used for C++

**Comment by andreeis** [comment]
@Randshot, @minhducsun2002, @ProfFan, did you have the same llvm installed for a while and the problem started to manifest with a CMake Tools update (like 1.7.0 or 1.7.1) or with an LLVM update?

@bobbrow, is it possible that your fix for https://github.com/microsoft/vscode-cmake-tools/issues/1821 to address this as well?

**Comment by bobbrow** [comment]
The original report in this issue is for version 1.6.0 and #1821 is a 1.7.0 regression, so I'm not sure they're related.  But folks on this thread can try installing my patch here to see if it fixes the problem.
1. Download from here: https://github.com/microsoft/vscode-cmake-tools/suites/2617691844/artifacts/57356650
2. Unzip the VSIX.
3. Use the "Install from VSIX" command in VS Code to install it.
4. Run the "Reload Window" command to ensure that it was installed.
5. Run the "Scan For Kits" command again to see if it fixed the problem.

**Comment by minhducsun2002** [comment]
> did you have the same llvm installed for a while and the problem started to manifest with a CMake Tools update (like 1.7.0 or 1.7.1) or with an LLVM update?

I did not use CMake Tools before 1.7.1, and have not updated LLVM at any point during usage of the extension.

**Comment by riri** [comment]
@bobbrow your fix for #1821 does not fix this issue, g++ is still used instead of clang++ (and IntellisenseMode switches from "linux-clang-x64" to "linux-gcc-x64").

I can do tests similars to those I did for the other bug if needed.

**Comment by andreeis** [comment]
@riri, did this always happen for you or it started when you upgraded the extension to a particular recent version or when you upgraded your llvm?

**Comment by riri** [comment]
@andreeis  I cannot tell really, I didn't start a long time ago to handle all this, before I didn't check if clang was properly configured.

**Comment by riri** [comment]
That's what @bobbrow 's fix (version `v1.7.2-63ccbad`) outputs for a kit scan with loggingLevel = debug:
```
[extension] [5441] cmake.scanForKits started
[kitsController] Rescanning for kits
[kit] Scanning for Kits on system
[kit] Scanning directory /home/riri/Work/go/bin for compilers
[kit] Scanning directory /home/riri/.local/bin for compilers
[kit] Scanning directory /usr/local/sbin for compilers
[kit] Scanning directory /usr/bin for compilers
[kit] Scanning directory /usr/local/bin for compilers
[kit] Scanning directory /usr/lib/jvm/default/bin for compilers
[kit] Scanning directory /usr/bin/site_perl for compilers
[kit] Scanning directory /usr/bin/vendor_perl for compilers
[kit] Scanning directory /usr/bin/core_perl for compilers
[kit] Skipping scan of not existing path undefined\LLVM\bin
[kit] Found 0 kits in directory undefined\LLVM\bin
[kit] Found 0 kits in directory /home/riri/Work/go/bin
[kit] Found 0 kits in directory /usr/local/sbin
[kit] Found 0 kits in directory /home/riri/.local/bin
[kit] Found 0 kits in directory /usr/local/bin
[kit] Found 0 kits in directory /usr/bin/site_perl
[kit] Found 0 kits in directory /usr/lib/jvm/default/bin
[kit] Found 0 kits in directory /usr/bin/vendor_perl
[kit] Found 0 kits in directory /usr/bin/core_perl
[kit] Testing Clang binary: /usr/bin/clang
[kit] Testing Clang binary: /usr/bin/clang
[proc] Executing command: /usr/bin/clang -v
[kit] Testing Clang binary: /usr/bin/clang-11
[kit] Testing Clang binary: /usr/bin/clang-11
[proc] Executing command: /usr/bin/clang-11 -v
[kit] Testing GCC binary: /usr/bin/gcc
[kit] Testing GCC binary: /usr/bin/gcc
[proc] Executing command: /usr/bin/gcc -v
[kit] Testing GCC binary: /usr/bin/x86_64-pc-linux-gnu-gcc
[kit] Testing GCC binary: /usr/bin/x86_64-pc-linux-gnu-gcc
[proc] Executing command: /usr/bin/x86_64-pc-linux-gnu-gcc -v
[kit] Testing GCC binary: /usr/bin/x86_64-pc-linux-gnu-gcc-10.2.0
[kit] Testing GCC binary: /usr/bin/x86_64-pc-linux-gnu-gcc-10.2.0
[proc] Executing command: /usr/bin/x86_64-pc-linux-gnu-gcc-10.2.0 -v
[kit] [GCC] Using built-in specs.
[kit] [GCC] COLLECT_GCC=/usr/bin/x86_64-pc-linux-gnu-gcc-10.2.0
[kit] [GCC] COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/lto-wrapper
[kit] [GCC] Target: x86_64-pc-linux-gnu
[kit] [GCC] Configured with: /build/gcc/src/gcc/configure --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --enable-languages=c,c++,ada,fortran,go,lto,objc,obj-c++,d --with-isl --with-linker-hash-style=gnu --with-system-zlib --enable-__cxa_atexit --enable-cet=auto --enable-checking=release --enable-clocale=gnu --enable-default-pie --enable-default-ssp --enable-gnu-indirect-function --enable-gnu-unique-object --enable-install-libiberty --enable-linker-build-id --enable-lto --enable-multilib --enable-plugin --enable-shared --enable-threads=posix --disable-libssp --disable-libstdcxx-pch --disable-libunwind-exceptions --disable-werror gdc_include_dir=/usr/include/dlang/gdc
[kit] [GCC] Thread model: posix
[kit] [GCC] Supported LTO compression algorithms: zlib zstd
[kit] [GCC] gcc version 10.2.0 (GCC)
[kit] Detected GCC compiler: /usr/bin/x86_64-pc-linux-gnu-gcc-10.2.0
[kit] [GCC] Using built-in specs.
[kit] [GCC] COLLECT_GCC=/usr/bin/x86_64-pc-linux-gnu-gcc
[kit] [GCC] COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/lto-wrapper
[kit] [GCC] Target: x86_64-pc-linux-gnu
[kit] [GCC] Configured with: /build/gcc/src/gcc/configure --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --enable-languages=c,c++,ada,fortran,go,lto,objc,obj-c++,d --with-isl --with-linker-hash-style=gnu --with-system-zlib --enable-__cxa_atexit --enable-cet=auto --enable-checking=release --enable-clocale=gnu --enable-default-pie --enable-default-ssp --enable-gnu-indirect-function --enable-gnu-unique-object --enable-install-libiberty --enable-linker-build-id --enable-lto --enable-multilib --enable-plugin --enable-shared --enable-threads=posix --disable-libssp --disable-libstdcxx-pch --disable-libunwind-exceptions --disable-werror gdc_include_dir=/usr/include/dlang/gdc
[kit] [GCC] Thread model: posix
[kit] [GCC] Supported LTO compression algorithms: zlib zstd
[kit] [GCC] gcc version 10.2.0 (GCC)
[kit] Detected GCC compiler: /usr/bin/x86_64-pc-linux-gnu-gcc
[kit] [GCC] Using built-in specs.
[kit] [GCC] COLLECT_GCC=/usr/bin/gcc
[kit] [GCC] COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/lto-wrapper
[kit] [GCC] Target: x86_64-pc-linux-gnu
[kit] [GCC] Configured with: /build/gcc/src/gcc/configure --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --enable-languages=c,c++,ada,fortran,go,lto,objc,obj-c++,d --with-isl --with-linker-hash-style=gnu --with-system-zlib --enable-__cxa_atexit --enable-cet=auto --enable-checking=release --enable-clocale=gnu --enable-default-pie --enable-default-ssp --enable-gnu-indirect-function --enable-gnu-unique-object --enable-install-libiberty --enable-linker-build-id --enable-lto --enable-multilib --enable-plugin --enable-shared --enable-threads=posix --disable-libssp --disable-libstdcxx-pch --disable-libunwind-exceptions --disable-werror gdc_include_dir=/usr/include/dlang/gdc
[kit] [GCC] Thread model: posix
[kit] [GCC] Supported LTO compression algorithms: zlib zstd
[kit] [GCC] gcc version 10.2.0 (GCC)
[kit] Detected GCC compiler: /usr/bin/gcc
[kit] [Clang] clang version 11.1.0
[kit] [Clang] Target: x86_64-pc-linux-gnu
[kit] [Clang] Thread model: posix
[kit] [Clang] InstalledDir: /usr/bin
[kit] [Clang] Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-pc-linux-gnu/10.2.0
[kit] [Clang] Found candidate GCC installation: /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/10.2.0
[kit] [Clang] Found candidate GCC installation: /usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0
[kit] [Clang] Found candidate GCC installation: /usr/lib64/gcc/x86_64-pc-linux-gnu/10.2.0
[kit] [Clang] Selected GCC installation: /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/10.2.0
[kit] [Clang] Candidate multilib: .;@m64
[kit] [Clang] Candidate multilib: 32;@m32
[kit] [Clang] Selected multilib: .;@m64
[kit] Detected Clang compiler: /usr/bin/clang-11
[kit] Detected Clang compiler: /usr/bin/clang-11
[kit] [Clang] clang version 11.1.0
[kit] [Clang] Target: x86_64-pc-linux-gnu
[kit] [Clang] Thread model: posix
[kit] [Clang] InstalledDir: /usr/bin
[kit] [Clang] Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-pc-linux-gnu/10.2.0
[kit] [Clang] Found candidate GCC installation: /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/10.2.0
[kit] [Clang] Found candidate GCC installation: /usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0
[kit] [Clang] Found candidate GCC installation: /usr/lib64/gcc/x86_64-pc-linux-gnu/10.2.0
[kit] [Clang] Selected GCC installation: /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/10.2.0
[kit] [Clang] Candidate multilib: .;@m64
[kit] [Clang] Candidate multilib: 32;@m32
[kit] [Clang] Selected multilib: .;@m64
[kit] Detected Clang compiler: /usr/bin/clang
[kit] Detected Clang compiler: /usr/bin/clang
[kit] Found 5 kits in directory /usr/bin
[kit] Found Kit: Clang 11.1.0 x86_64-pc-linux-gnu
[kit] Found Kit: Clang 11.1.0 x86_64-pc-linux-gnu
[kit] Found Kit: GCC 10.2.0 x86_64-pc-linux-gnu
[kit] Found Kit: GCC 10.2.0 x86_64-pc-linux-gnu
[kit] Found Kit: GCC 10.2.0 x86_64-pc-linux-gnu
[kitsController] Saving kits to /home/riri/.local/share/CMakeTools/cmake-tools-kits.json
[kitsController] Saving new kits to /home/riri/.local/share/CMakeTools/cmake-tools-kits.json
[extension] [5441] cmake.scanForKits finished (returned undefined)
[kit] Reading kits file /home/riri/.local/share/CMakeTools/cmake-tools-kits.json
[kit] Successfully loaded 2 kits from /home/riri/.local/share/CMakeTools/cmake-tools-kits.json
```

**Comment by ProfFan** [comment]
Note since Clang 11 is very new it's unlikely that anyone would see this issue happening with older extension (since we almost always autoupdate)

**Comment by andreeis** [comment]
I installed clang 11.0 on an ubuntu VM and scan for kits finds it with C and CXX compilers. I'll install 11.1 exactly to see what happens.

**Comment by ProfFan** [comment]
@andreeis If there is a debug output mode on the kit scanner I can provide you logs :)

Or I can try use the DevTools

**Comment by andreeis** [comment]
Unfortunately, the logging for the kits scanning feature is not very useful. We will improve it and once the change is merged in the code base we will create a vsix for you to reproduce this issue again and send us better logs.
Until then, you can complete your kit definition with the CXX compiler and make sure you don't scan again (it will overwrite your edit). Let us know if you encounter any issues with fixing the kit definition directly in the kits file.

**Comment by riri** [comment]
It's ok, I have manually changed the c++ definition for the clang kit.
Thank you for the support, I will test with pleasure when a fix is ready to test.

**Comment by ProfFan** [comment]
@andreeis Hi, please ping me when the logging for kit scanning is ready, so I can help test the configurations :)

**Comment by waterlens** [comment]
Has this problem been solved?

**Comment by bobbrow** [comment]
The fix for this issue is available in 1.10 "pre-release".  You can help us validate it by:
1. Opening the extensions panel in VS Code
2. Selecting the CMake Tools extension
3. Clicking the "Switch to Pre-release Version" button

![image](https://user-images.githubusercontent.com/12818240/156218538-7bce4a1f-4fc9-43cc-afb0-7f423aa5980e.png)
